Add start time play option

aprotyas opened the pull request:
This PR adds a `--start-time` (optional) argument to the `ros2 bag play` verb. The desired use is as follows:

```python
ros2 bag play --start-time [positive-float-denoting-start-time-in-seconds]
```

In doing so, the `start_offset` field is added to `rosbag2_transport::PlayOptions` class and its corresponding PyBind11 wrapper.

`rosbag2_transport::Player` respects this option by adding `start_offset` - during construction - to the `starting_time` obtained from its readers metadata, but only does so if the start time is a valid non-negative value. Else, `Player` outputs to the warning stream about the invalid offset.

Partially addresses https://github.com/ros2/rosbag2/issues/929 and gets closer to parity with `rosbag play` from ROS 1.

Signed-off-by: Abrar Rahman Protyasha <aprotyas@u.rochester.edu>

aprotyas commented:
Note for maintainers: this is still in draft because I'd like to get some preliminary feedback on the implementation. The main concern revolves around supplanting `starting_time_` in `Player::play()`. My rationale behind doing that is - if the `starting_time` provided through `PlayOptions` is to be respected, it must also serve as the base point in the `Player::seek()` method. As the `starting_time_` field sees no other use within the `Player` class, this should (?) be safe.

aprotyas commented:
CI:
Repos file: https://gist.githubusercontent.com/aprotyas/caee0f1688ca30d6f8c9abb787b7ec3d/raw/9c3e6b6b69ec8372c791ab9bff0573fd64ff063f/ros2.repos
Build args: `--packages-up-to ros2bag`
Test args: `--packages-select ros2bag rosbag2_py rosbag2_transport`
Job: https://ci.ros2.org/job/ci_launcher/9446/

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=15720)](http://ci.ros2.org/job/ci_linux/15720/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=10402)](http://ci.ros2.org/job/ci_linux-aarch64/10402/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=13385)](http://ci.ros2.org/job/ci_osx/13385/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=15964)](http://ci.ros2.org/job/ci_windows/15964/)

gbiggs commented on rosbag2_transport/test/rosbag2_transport/test_play_timing.cpp:
```diff
@@ -190,3 +191,44 @@ TEST_F(PlayerTestFixture, play_ignores_invalid_delay)
+  EXPECT_THAT(replay_time, Lt(upper_expected_duration));
+}
+
+TEST_F(PlayerTestFixture, play_ignores_invalid_starting_time)
```
There should be some kind of warning to the user when the starting time is invalid.

aprotyas replied:
That makes sense - I guess I overlooked it because the CLI invocation doesn't allow an invalid starting time, but that doesn't stop someone from populating a `rosbag2_transport::PlayOptions` object with the wrong value. Where should this invariant be exercised? It's convenient (not so for the user) to enforce the invariant lazily - i.e. during a `seek`/`play` function - but that may not be the greatest design.

aprotyas replied:
I guess if we didn't want to push the negativity check down to the player, then `rosbag2_transport::PlayOptions::starting_time` could be wrapped as a `non_negative<rcutils_time_point_value_t>` with an `operator rcutils_time_point_value_t()` implementation for implicit conversions between the wrapper and the underlying integer type - wrapper in this gist:

[Non-negative signed integral type wrapper](https://gist.github.com/aprotyas/02803a4ade50059285ec8f7badbf2edd)

gbiggs replied:
I like the concept that the player options must be correct as constructed, and making the `starting_time` a `non_negative<T>` value allows that - it would cause an error rapidly for the user. However if it's difficult to integrate that cleanly then I'd be happy with checking the options struct's values at the point of use and throwing an error there.

aprotyas replied:
In hindsight, it may be hard to integrate this `non_negative<T>`, if for nothing else but for its "sole occurrence" nature in the `PlayOptions` class.
The warning/error for a negative `starting_time` definitely has to go somewhere, but its better to first sort out how the two `starting_time` sources - one from `PlayOptions` and one from `BagMetadata` - live together.

emersonknapp replied:
I think it makes OK sense to validate `start_offset` in `Player` usage. As it is semantically a Duration, in my view there's nothing wrong theoretically with the data having a negative duration. However, we would consider it a constraint of this specific `Player` implementation that it does not want to start playback before the beginning of the bag (though it would have no technical problem doing so, this use case feels like it's probably not something we would actually want to do in practice).

We're already doing validation of `delay` in `Player::play`, I wouldn't argue against validating `start_offset` there as well.

aprotyas replied:
So, as of https://github.com/ros2/rosbag2/pull/931/commits/c79d65cfacb565d6c8b7829ba2bdefce5d61cac0, if the `start_offset` is a negative value, `Player` outputs a message to the warning stream informing the user that it's invalid, but otherwise just ignores the negative value. Thoughts?

emersonknapp commented on rosbag2_transport/include/rosbag2_transport/play_options.hpp:
```diff
@@ -53,6 +54,9 @@ struct PlayOptions
   bool start_paused = false;
 
+  // Start time from which the player should play the bag.
+  rcutils_time_point_value_t starting_time = INT64_C(0);
```
nitpick: I would prefer to call this `start_time`. Or maybe even `start_offset`

The docstring should definitely be specific that this is an offset. "Time to start  playback as an offset from the beginning of the bag"

aprotyas replied:
To be honest, `start_time` does not do justice to the difference in meaning between this field and `starting_time` from `BagMetadata::starting_time`. I'm inclined to call it something like `start_offset` or `start_from` instead - probably the former.

I'll take extra care to note its nature in documentation.

aprotyas replied:
Renamed in https://github.com/ros2/rosbag2/pull/931/commits/ec56a7a01c7ec726f9f3744ceaff410bd14171a4.

emersonknapp commented on rosbag2_transport/src/rosbag2_transport/player.cpp:
```diff
@@ -191,6 +191,12 @@ void Player::play()
+  // If a non-default (positive) starting time is provided in PlayOptions,
+  // override the starting time obtained from reader metadata
+  if (play_options_.starting_time > INT64_C(0)) {
+    starting_time_ = play_options_.starting_time;
```
> how - in general - PlayOptions::starting_time and BagMetadata::starting_time reside together - do they represent the same information?

I think that this option makes most sense as an "offset from the starting_time of the bag". `BagMetadata::starting_time` means "the timestamp of the first message in the bag" (though there is some discussion of changing it to mean "the timestamp at the time recording was started" - that is not the meaning presently). This value can be anything, as it depends on the recording system's steady clock.

However, the user likely does not care what the "absolute time" for starting is. E.g. if a bag has `starting_time=100`, and the user wants to start 20 seconds into the bag, they probably want to say `--start-time 20`, not `--start-time 120`, since the latter requires extra knowledge about the specific bag.

Given that, I would be more in favor this line looking like
```suggestion
    starting_time_ = metadata.starting_time + play_options_.start_offset;
```

aprotyas replied:
> I think that this option makes most sense as an "offset from the starting_time of the bag". 

Yes, that's exactly how I expect this to be used. I concur that the user would mean play from 20 seconds into the bag when they will use `--start-time 20`, so I will incorporate this suggested change.

On a related note, where should error checking for the `play_options_.starting_time` go? Does it make sense to throw a runtime error if `play_options_.starting_time < 0` here? Or should an error/warning be raised at construction/assignment time? Check https://github.com/ros2/rosbag2/pull/931#discussion_r763526466 for more context.

aprotyas replied:
I added the `start_offset` to `starting_time` in the `Player` constructor rather than the `play` function - primarily because that means we only add to the `starting_time` only once. From that point onwards, it doesn't matter if an offset was specified in `play_options_`, I think.

Check https://github.com/ros2/rosbag2/pull/931/commits/c79d65cfacb565d6c8b7829ba2bdefce5d61cac0 for reference.

emersonknapp reviewed (changes requested):
Left a longer comment inline - lmk what you think about that

aprotyas commented:
Running CI after addressing feedback:
Repos file: https://gist.githubusercontent.com/aprotyas/caee0f1688ca30d6f8c9abb787b7ec3d/raw/9c3e6b6b69ec8372c791ab9bff0573fd64ff063f/ros2.repos
Build args: `--packages-up-to ros2bag`
Test args: `--packages-select ros2bag rosbag2_py rosbag2_transport`
Job: https://ci.ros2.org/job/ci_launcher/9469/

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=15743)](http://ci.ros2.org/job/ci_linux/15743/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=10425)](http://ci.ros2.org/job/ci_linux-aarch64/10425/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=13408)](http://ci.ros2.org/job/ci_osx/13408/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=15990)](http://ci.ros2.org/job/ci_windows/15990/)

MichaelOrlov commented on rosbag2_transport/include/rosbag2_transport/play_options.hpp:
```diff
@@ -53,6 +54,9 @@ struct PlayOptions
   bool start_paused = false;
 
+  // Time to start playback as an offset from the beginning of the bag.
+  rcutils_time_point_value_t start_offset = INT64_C(0);
```
```suggestion
  rcutils_time_point_value_t start_offset = 0;
```
There are no reason for this ridiculous type cast.

aprotyas replied:
Noted - my bad! I definitely over-engineered things here. :sweat_smile: 

Addressed in https://github.com/ros2/rosbag2/pull/931/commits/07543e0c770db06fbada46268105fd4eea1a9d3a.

MichaelOrlov commented on rosbag2_transport/src/rosbag2_transport/player.cpp:
```diff
@@ -142,6 +142,17 @@ Player::Player(
       metadata.starting_time.time_since_epoch()).count();
+    // If a non-default (positive) starting time offset is provided in PlayOptions,
+    // then add the offset to the starting time obtained from reader metadata
+    if (play_options_.start_offset < INT64_C(0)) {
```
```suggestion
    if (play_options_.start_offset < 0) {
```

aprotyas replied:
Addressed in https://github.com/ros2/rosbag2/pull/931/commits/07543e0c770db06fbada46268105fd4eea1a9d3a.

MichaelOrlov commented on rosbag2_transport/test/rosbag2_transport/test_play_timing.cpp:
```diff
@@ -190,3 +191,43 @@ TEST_F(PlayerTestFixture, play_ignores_invalid_delay)
+
+TEST_F(PlayerTestFixture, play_respects_start_offset)
+{
+  rclcpp::Duration start_delay(0, static_cast<uint32_t>(5e8));  // 5e8 ns == 0.5 s
```
```suggestion
  rclcpp::Duration start_delay(0, static_cast<uint32_t>(RCUTILS_S_TO_NS(0.5)));
```

aprotyas replied:
Addressed in https://github.com/ros2/rosbag2/pull/931/commits/07543e0c770db06fbada46268105fd4eea1a9d3a.

MichaelOrlov commented on rosbag2_transport/test/rosbag2_transport/test_play_timing.cpp:
```diff
@@ -190,3 +191,43 @@ TEST_F(PlayerTestFixture, play_ignores_invalid_delay)
+  rclcpp::Duration delay_margin(1, 0);
+
+  // Player should ignore an invalid (negative) start offset
+  play_options_.start_offset = INT64_C(-5);
```
```suggestion
  play_options_.start_offset = -5;
```

aprotyas replied:
Addressed in https://github.com/ros2/rosbag2/pull/931/commits/07543e0c770db06fbada46268105fd4eea1a9d3a.

MichaelOrlov commented on rosbag2_transport/include/rosbag2_transport/play_options.hpp:
```diff
@@ -16,6 +16,7 @@
 #define ROSBAG2_TRANSPORT__PLAY_OPTIONS_HPP_
 
 #include <cstddef>
+#include <cstdint>
```
```suggestion
```

aprotyas replied:
Addressed in https://github.com/ros2/rosbag2/pull/931/commits/07543e0c770db06fbada46268105fd4eea1a9d3a.

MichaelOrlov commented on ros2bag/ros2bag/verb/play.py:
```diff
@@ -90,6 +90,9 @@ def add_arguments(self, parser, cli_name):  # noqa: D102
             '-p', '--start-paused', action='store_true', default=False,
             help='Start the playback player in a paused state.')
+        parser.add_argument(
+            '--start-time', type=check_positive_float, default=0.0,
```
Naming parameter as `--start-time` and mapping it internally to the `start_offset` causing a lot of confusion what it really is.
It could be not clear that `start-time` doesn't really means start time in rosbag timestamp but rather time offset from the begging of the bag.
IMO it would be better to rename parameter `--start-time` to the `--start-offset`.

aprotyas replied:
Yes, I agree that `--start-offset` would be less confusing for developers. The reason I initially picked `--start-time` for the CLI argument was because I thought the wording would align with what (most) users expected - time offset from the beginning of the bag.

Having said that, it's better to not leave room for ambiguity. Addressed in https://github.com/ros2/rosbag2/pull/931/commits/975c65c9411d028cfaf7d140423c00e9679306a8.

jhdcs commented on rosbag2_transport/test/rosbag2_transport/test_play_timing.cpp:
```diff
@@ -190,3 +190,43 @@ TEST_F(PlayerTestFixture, play_ignores_invalid_delay)
+
+  auto start = clock.now();
+  player->play();
+  auto replay_time = clock.now() - start;
```
Nitpick: Would this be slightly clearer if named `replay_duration`?

aprotyas replied:
Agreed! Renamed in https://github.com/ros2/rosbag2/pull/931/commits/7cd158f6163cc7f2538140a1528645669c55f641

jhdcs reviewed (commented):
Just a minor naming thing to add, probably a much lower priority that the others' comments/suggestions.

emersonknapp reviewed (approved):
This LGTM with the updates. Thanks for the revision

aprotyas commented:
Thank you for the reviews everyone. Running CI again after the latest round of revisions:

Repos file: https://gist.githubusercontent.com/aprotyas/caee0f1688ca30d6f8c9abb787b7ec3d/raw/9c3e6b6b69ec8372c791ab9bff0573fd64ff063f/ros2.repos
Build args: `--packages-up-to ros2bag`
Test args: `--packages-select ros2bag rosbag2_py rosbag2_transport`
Job: https://ci.ros2.org/job/ci_launcher/9475/

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=15750)](http://ci.ros2.org/job/ci_linux/15750/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=10431)](http://ci.ros2.org/job/ci_linux-aarch64/10431/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=13415)](http://ci.ros2.org/job/ci_osx/13415/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=16001)](http://ci.ros2.org/job/ci_windows/16001/)

aprotyas commented:
@emersonknapp I was wondering whether PRs which modify existing or add new features like this should be documented in the Humble release notes? Is that effort being tracked separately? I ask because I don't see any rosbag2 related entries in the release notes: https://docs.ros.org/en/rolling/Releases/Release-Humble-Hawksbill.html#changes-since-the-galactic-release

adiser commented:
Is there a backport of this feature planned for Foxy?

MichaelOrlov commented:
@adiser This PR will not be backported to the Foxy at least because it has breaking ABI changes.
